### ISSUE TITLE
compile error on OS X with gcc or icc

### ISSUE THREAD

**Issue by zzlei** [magnumripper/JohnTheRipper#1203]
**with gcc:**
gcc-4.9 -DAC_BUILT  -DARCH_LITTLE_ENDIAN=1 -march=native -mavx -c -fopenmp   -DUNDERSCORES -DBSD -DALIGN_LOG x86-64.S
x86-64.S:1651:no such instruction: `xgetbv'

**with icc:**
icc -DAC_BUILT -march=native -mavx -c -g -O2 -I/usr/local/include -DARCH_LITTLE_ENDIAN=1     -fopenmp  -D_THREAD_SAFE -pthread  -I/usr/local/include  listconf.c -o listconf.o
icc: command line warning #10120: overriding '-march=native' with '-mavx'
listconf.c(193): error: extra text after expected end of number
  	printf("clang version: %s\n", \__clang_version__);
  	                              ^

The latter seems to be a bug in icc. See this [post](https://software.intel.com/en-us/forums/topic/549663).

Specs:
OS X 10.10.3
gcc 4.9.2
icc 15.0.2

**Comment by magnumripper** [comment]
I use gcc 4.9.2 (from Homebrew) with no problems. OSX native 'ld' doesn't support AVX so you need to pass `-Wa,-q` to gcc to make it end up in using clang for linking. Our autoconf takes care of that.

**Comment by magnumripper** [comment]
Sorry I meant 'as' not 'ld'

**Comment by magnumripper** [comment]
For icc, try this
```
icc -dM -E -x c /dev/null | grep clang
```

**Comment by magnumripper** [comment]
```
$ clang -dM -E -x c /dev/null | grep clang_version
#define __clang_version__ "6.1.0 (clang-602.0.49)"
```
Perhaps icc defines a string for some (dumb) compatibility reason but omits the quotes?

**Comment by magnumripper** [comment]
Whenever closing an issue, please add some comment so we know why or what was the solution! Did the `-Wa,-q` do the trick for gcc? Did you find a workaround for icc? I just tried my mentioned macro output command with icc on super, and it did not show anything at all matching clang. Though I guess that version doesn't have the problem.

**Comment by zzlei** [comment]
I'm sorry...

The -Wa,-q flags work well for gcc. I think the problem with icc is specific on OS X, where clang is the default compiler.d

**Comment by magnumripper** [comment]
> I think the problem with icc is specific on OS X, where clang is the default compiler.

We still want to work around it. Perhaps we could do something like
```
#if __clang_version__ && !__INTEL_COMPILER
```

**Comment by zzlei** [comment]
Well, gcc still has trouble compiling JtR, when OpenCL is enabled:
```
gcc-4.9 -Wa,-q -DAC_BUILT -march=native -mavx -c -g -O2 -I/usr/local/include -DARCH_LITTLE_ENDIAN=1   -Wall -Wdeclaration-after-statement -fomit-frame-pointer --param allow-store-data-races=0 -Wno-deprecated-declarations -Wno-format-extra-args -std=gnu89    -fopenmp  -D_THREAD_SAFE -pthread  -I/usr/local/include -DHAVE_OPENCL -D_THREAD_SAFE -pthread -funroll-loops common_opencl_pbkdf2_plug.c -o common_opencl_pbkdf2_plug.o
In file included from /usr/include/dispatch/dispatch.h:51:0,
                 from /System/Library/Frameworks/OpenCL.framework/Headers/gcl.h:23,
                 from /System/Library/Frameworks/OpenCL.framework/Headers/opencl.h:16,
                 from common-opencl.h:19,
                 from common_opencl_pbkdf2.h:11,
                 from common_opencl_pbkdf2_plug.c:13:
/usr/include/dispatch/object.h:143:15: error: expected identifier or '(' before '^' token
 typedef void (^dispatch_block_t)(void);
               ^
/usr/include/dispatch/object.h:362:3: error: unknown type name 'dispatch_block_t'
   dispatch_block_t notification_block);
   ^
```

**Comment by magnumripper** [comment]
This is again described in doc/INSTALL, last section. For some odd reason you need to patch the system headers. Backups will be taken. The changes are reverted at each OS update so you need to re-apply.

Executive summary:
```
cd john/src
$ sudo patch -b -N -p0 < unused/Yosemite.patch
```

**Comment by zzlei** [comment]
Alright. I should've read those docs more thoroughly.

**Comment by magnumripper** [comment]
I'm actually not quite sure whether it's a bug in gcc or in the headers. That patch is not needed for clang.

**Comment by magnumripper** [comment]
BTW I found a problem in autoconf when forcing SSE4.1 on an AVX(+) host:
```
./configure CC="gcc -msse4.1" --disable-native-tests
```
It would end up as AVX anyway and then fail with that "xgetbv" error for some odd reason. Commited a fix to bleeding-jumbo only (95e5305), I don't want a commit pushed to both branches.
